lib/Net/Stripe.pm

```diff
@@ -147,6 +147,16 @@ if your account is using Stripe API version 2015-02-18 or later, you should
 update any code that relies on strict checking of the return value of
 Charge->status.
 
+=item deprecate direct handling of PANs
+
+Stripe strongly discourages direct handling of PANs (primary account numbers),
+even in test mode, and returns invalid_request_error when passing PANs to the
+API from accounts that were created after October 2017. In live mode, all
+tokenization should be performed via client-side libraries because direct
+handling of PANs violates PCI compliance. So we have removed the methods and
+parameter constraints that allow direct handling of PANs and updated the
+unit tests appropriately.
```

Comment:
I'm not sure about this. If Stripe are still supporting it for pre-October 2017 accounts, shouldn't Net::Stripe keep supporting it?

You could add noisy warnings so that people know they're doing something wrong and then come and read this.

Reply:
i am of a few minds about it:
* while it is not our business how users want to configure and use their Stripe accounts, if we continue to support PAN handling, then we must have a dependable way of exercising those portions of the SDK after making updates that may impact them.
* any future maintainer or contributor with a newer account, or users installing from CPAN with an `$ENV{STRIPE_API_KEY}` set, will fail those portions of the test suite that test PAN handling, unless they modify their account settings.
 * conversely, if we modify the test script to expect or swallow errors for these unit tests, then it seems that we are effectively not testing the PAN-handling potions of the SDK.
 * from the Stripe perspective, technically it is not a matter of support vs not support. if you so desire, you can allow direct PAN handling for any account by enabling "Handle card information directly"
[here](https://dashboard.stripe.com/account/integration/settings). of course that setting is account-wide, allowing PAN handling in live mode as well.
 * i did some poking around in the stripe-supported SDKs and could not easily find an instance where they are facilitating PAN handling.

sooooooo... per the changes in this branch, there are a lot of tests that we can easily modify to handle test tokens instead of card numbers, without impacting their utility. but the sum total of the above is what moved me from my initial position of removing the unit tests that would fail without this account setting, to full deprecation.

of course i tagged you as a reviewer in hopes that you would provide some useful counter-thought, and i would like to hear any thoughts you have about how we might restructure the unit tests so that maintainers or contributors with modern accounts can still effectively test PAN-handling code.

Reply:
Thanks for taking so much care on this decision making process Sherrard. From everything you've told me above, here are my thoughts about a user handling PANs.

- If they're not aware of the ramifications, this change will ensure that they are.
- If they're unable to change the process due to the complexity of their own systems, they have the option to roll back to earlier releases of Net::Stripe or write their own patch.

For this reason I think you should go ahead with removing PAN handling from Net::Stripe.

Reply:
agreed.

i also recognize a third option, which is
 * there is a legitimate, non-PAN use case for one of the methods or constraints that are being removed

if so, we can always make future adjustments armed with actual use cases.